charts/turing/templates/_helpers.tpl

```diff
@@ -178,8 +186,19 @@ API config related
 {{- $globMlpApiHost := include "turing.get-workload-host" (list .Values.global .Release.Namespace "mlp")}}
 AuthConfig:
   URL: {{ include "turing.authorization.server.url" . | quote }}
+{{ if and (eq (toString .Values.config.BatchEnsemblingConfig.Enabled) "true") (.Values.imageBuilder.serviceAccount.name)}}
```

Comment:
For the checks here, shouldn't we be checking if either `(.Values.imageBuilder.serviceAccount.name)` is set or `(.Values.imageBuilder.serviceAccount.create)` is `true`? Same for the check below, under EnsemblerServiceBuilderConfig. This being the case, shall we make the SA name in the values "" by default?

```
imageBuilder:
  serviceAccount:
    create: true
    name: ""
    annotations: {}
    labels: {}
```

Reply:
Oh actually it's not the service account details that we're concerned about (yes the default values provide a value for them so those are taken care of). Rather, it's the other fields in the `KanikoConfig` block expected by the API server (see https://github.com/caraml-dev/turing/blob/main/api/turing/config/config.go#L197) **that would be missing in the `BatchEnsemblingConfig` block** if we set the service account field details without them (see https://github.com/caraml-dev/helm-charts/actions/runs/5819088465/job/15776808845#step:12:295). 

Essentially, the `.Values.imageBuilder.serviceAccount` values get propagated to two different `KanikoConfig` blocks, one within the `BatchEnsemblingConfig` block and the other within the `EnsemblerServiceBuilderConfig` [block](https://github.com/caraml-dev/helm-charts/blob/main/charts/turing/values.yaml#L180). As it's possible to turn off batch ensembling but not the pyfunc ensembler service workflows, it's also normal to expect the batch ensembling chart values to not be populated (just as in the default values.yaml file). If we were to inject the service account details into the batch ensembling block even though it's turned off, the Go validator in the Turing API server would end up validating the entire `BatchEnsemblingConfig.ImageBuildingConfig` and `BatchEnsemblingConfig.ImageBuildingConfig.KanikoConfig` blocks without the other fields set and throw the errors linked above. I'm still not sure why the field's being validated by the validator since we've set a conditional [tag](https://github.com/caraml-dev/turing/blob/main/api/turing/config/config.go#L124) on it 🤔 Maybe there's something wrong with the library or the we're using the tags 😐

Not sure if I made sense or if I'm even answering your question 😅 Let me try to sit back and figure out what you were trying to suggest from your point of view.

Reply:
> For the checks here, shouldn't we be checking if either (.Values.imageBuilder.serviceAccount.name) is set or (.Values.imageBuilder.serviceAccount.create) is true? 

Hmm okay, but what should the expected behaviour be when `.Values.imageBuilder.serviceAccount.name` is not set but `.Values.imageBuilder.serviceAccount.create` is? Should we be creating a service account with a default name and then injecting it into the API server configs? Seems like in Merlin we're doing the same checks (I just copied the Kaniko stuff from the Merlin chart). 

But yes I agree that the intention behind the check you suggested makes sense (it doesn't make sense to create a service account but not use it anywhere if the name specified is ""). In this case, it probably is more logical to perform the `(.Values.imageBuilder.serviceAccount.create)` check.

Reply:
Actually my question is not about whether we should check `.Values.config.BatchEnsemblingConfig.Enabled` - it's clear that it must be checked. It's more about why we are only checking `.Values.imageBuilder.serviceAccount.name` and not `.Values.imageBuilder.serviceAccount.create` as well.

> Should we be creating a service account with the default name and then inject it into the API server configs i.e. when .Values.imageBuilder.serviceAccount.name == "", we create a service account with a default name and then add it into the configs?

This was my understanding from the implementation of the helper function `{{- define "turing.kaniko-sa" -}}`. The helper function caters to `.Values.imageBuilder.serviceAccount.name` being "" by applying "kaniko" as the default. So, for completeness, I guess we can check if either `.Values.imageBuilder.serviceAccount.name` is set or `.Values.imageBuilder.serviceAccount.create` is true. (Probably Merlin needs the same change.) Wdyt?

Reply:
In this case, the helper function definition probably needs to change a little too (we only define `turing.kaniko-sa` as its default value if the kaniko service account name is not set, whether or not this value is used is depends on the individual sections that consume `turing.kaniko-sa`)

```tpl
{{- define "turing.kaniko-sa" -}}
{{- if .Values.imageBuilder.serviceAccount.name }}
{{- printf "%s" .Values.imageBuilder.serviceAccount.name }}
{{- else }}
{{- printf  "%s-%s" (default "kaniko" .Values.imageBuilder.serviceAccount.name) (include "turing.fullname" . ) }}
{{- end }}
{{- end }}
```